File: examples/node/client-sdk.mjs
/**
 * NVatar SDK — Node.js Client Example
 *
 * Connect to NVatar from your Node.js/Bun service.
 * Uses native fetch + WebSocket for zero dependencies.
 *
 * Usage:
 *   node client-sdk.mjs
 *   # or
 *   bun run client-sdk.mjs
 *
 * Requirements:
 *   Node.js 18+ (native fetch + WebSocket)
 *   or Bun (any version)
 */

const NVATAR_URL = process.env.NVATAR_URL || 'http://localhost:54444';
const WS_URL = NVATAR_URL.replace('https://', 'wss://').replace('http://', 'ws://');

// ---------------------------------------------------------------------------
// NVatar Client
// ---------------------------------------------------------------------------

class NVatarClient {
  constructor(baseUrl = NVATAR_URL) {
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.wsUrl = this.baseUrl.replace('https://', 'wss://').replace('http://', 'ws://');
  }

  // --- Avatar ---

  async getAvatar(avatarId) {
    const res = await fetch(`${this.baseUrl}/api/v1/avatars/${avatarId}`);
    if (!res.ok) throw new Error(`getAvatar failed: ${res.status}`);
    return res.json();
  }

  async createAvatar({ userId, name, persona, mbti, tone = '친근한', language = 'ko' }) {
    const res = await fetch(`${this.baseUrl}/api/v1/avatars`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        user_id: userId, name, persona, mbti, tone,
        speech_level: 'polite', language,
      }),
    });
    if (!res.ok) throw new Error(`createAvatar failed: ${res.status}`);
    return res.json();
  }

  // --- SDK Session ---

  async sdkConnect(avatarId, { saveFranchiseMemory = true } = {}) {
    const res = await fetch(`${this.baseUrl}/api/v1/sdk/connect`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        avatar_id: avatarId,
        save_franchise_memory: saveFranchiseMemory,
      }),
    });
    if (!res.ok) throw new Error(`sdkConnect failed: ${res.status}`);
    return res.json();
  }

  async sdkDisconnect(avatarId) {
    const res = await fetch(`${this.baseUrl}/api/v1/sdk/disconnect`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ avatar_id: avatarId }),
    });
    if (!res.ok) throw new Error(`sdkDisconnect failed: ${res.status}`);
    return res.json();
  }

  // --- Chat via WebSocket ---

  async chat(avatarId, messages, { onResponse, timeout = 15000 } = {}) {
    const ws = new WebSocket(`${this.wsUrl}/ws/chat/${avatarId}`);

    return new Promise((resolve, reject) => {
      const responses = [];

      ws.onmessage = (event) => {
        let data;
        try { data = JSON.parse(event.data); } catch { return; }
        const text = data.text || data.message || '';
        if (text && onResponse) {
          onResponse(data.type, text);
        }
        if (text) responses.push({ type: data.type, text });
      };

      ws.onopen = async () => {
        // Wait for initial messages (first meeting, proactive)
        await sleep(3000);

        for (const msg of messages) {
          ws.send(JSON.stringify({ type: 'message', text: msg }));
          await sleep(timeout);  // Wait for response
        }

        ws.close();
        resolve(responses);
      };

      ws.onerror = (err) => reject(err);
    });
  }
}

function sleep(ms) {
  return new Promise(r => setTimeout(r, ms));
}

// ---------------------------------------------------------------------------
// Example
// ---------------------------------------------------------------------------

async function main() {
  const client = new NVatarClient();

  // 1. Get avatar info
  const avatar = await client.getAvatar(155);
  const avatarData = avatar.response || avatar;
  console.log(`Avatar: ${avatarData.name || 'unknown'}`);

  // 2. Connect SDK
  const session = await client.sdkConnect(155, { saveFranchiseMemory: true });
  console.log('SDK connected:', session.ok);

  // 3. Chat
  const responses = await client.chat(155, [
    '안녕! 오늘 기분 어때?',
    '요즘 뭐 하고 지내?',
  ], {
    onResponse: (type, text) => {
      console.log(`  [${type}] ${text.slice(0, 100)}`);
    },
    timeout: 12000,
  });

  console.log(`\nTotal responses: ${responses.length}`);

  // 4. Disconnect
  await client.sdkDisconnect(155);
  console.log('Done!');
}

main().catch(console.error);

File: examples/python/client_sdk.py
"""
NVatar SDK — Python Client Example

Connect to NVatar from your own service.
This is for EXTERNAL integration — when your app calls NVatar APIs
rather than running code inside the NVatar server.

Use cases:
  - Mobile app connecting to NVatar avatar
  - Web service triggering avatar conversations
  - Backend service managing avatar sessions

Requirements:
  pip install httpx websockets
"""
import asyncio
import json
import httpx
import websockets


NVATAR_URL = "https://nvatar.nskit.io"  # or localhost:54444


class NVatarClient:
    """Simple client for NVatar SDK APIs."""

    def __init__(self, base_url: str = NVATAR_URL):
        self.base_url = base_url.rstrip("/")
        self.ws_url = self.base_url.replace("https://", "wss://").replace("http://", "ws://")

    # --- Avatar Management ---

    async def create_avatar(self, user_id: str, name: str, persona: str, **kwargs) -> dict:
        """Create a new avatar."""
        async with httpx.AsyncClient() as client:
            resp = await client.post(f"{self.base_url}/api/v1/avatars", json={
                "user_id": user_id,
                "name": name,
                "persona": persona,
                "mbti": kwargs.get("mbti"),
                "tone": kwargs.get("tone", "친근한"),
                "speech_level": kwargs.get("speech_level", "polite"),
                "language": kwargs.get("language", "ko"),
            })
            resp.raise_for_status()
            return resp.json()

    async def get_avatar(self, avatar_id: int) -> dict:
        """Get avatar details."""
        async with httpx.AsyncClient() as client:
            resp = await client.get(f"{self.base_url}/api/v1/avatars/{avatar_id}")
            resp.raise_for_status()
            return resp.json()

    # --- SDK Session ---

    async def sdk_connect(self, avatar_id: int, save_franchise_memory: bool = True) -> dict:
        """Initialize SDK session for an avatar."""
        async with httpx.AsyncClient() as client:
            resp = await client.post(f"{self.base_url}/api/v1/sdk/connect", json={
                "avatar_id": avatar_id,
                "save_franchise_memory": save_franchise_memory,
            })
            resp.raise_for_status()
            return resp.json()

    async def sdk_disconnect(self, avatar_id: int) -> dict:
        """End SDK session."""
        async with httpx.AsyncClient() as client:
            resp = await client.post(f"{self.base_url}/api/v1/sdk/disconnect", json={
                "avatar_id": avatar_id,
            })
            resp.raise_for_status()
            return resp.json()

    # --- Chat via WebSocket ---

    async def chat(self, avatar_id: int, messages: list[str], on_response=None):
        """Send messages and receive responses via WebSocket.

        Args:
            avatar_id: Avatar to chat with
            messages: List of user messages to send
            on_response: Callback for each response (type, text)
        """
        uri = f"{self.ws_url}/ws/chat/{avatar_id}"

        async with websockets.connect(uri) as ws:
            # Receive initial messages (first meeting, proactive)
            await self._drain(ws, on_response, timeout=3)

            for msg in messages:
                await ws.send(json.dumps({"type": "message", "text": msg}))
                await self._drain(ws, on_response, timeout=15)

    async def _drain(self, ws, callback, timeout: float = 5):
        """Drain WebSocket messages until timeout."""
        import time
        start = time.time()
        while time.time() - start < timeout:
            try:
                raw = await asyncio.wait_for(ws.recv(), timeout=2)
                try:
                    data = json.loads(raw)
                except json.JSONDecodeError:
                    continue
                msg_type = data.get("type", "")
                text = data.get("text", data.get("message", ""))
                if callback and text:
                    callback(msg_type, text)
            except asyncio.TimeoutError:
                continue


# ---------------------------------------------------------------------------
# Example Usage
# ---------------------------------------------------------------------------

async def main():
    client = NVatarClient("http://localhost:54444")

    # 1. Get avatar
    avatar = await client.get_avatar(155)
    avatar_data = avatar.get("response", avatar)
    print(f"Avatar: {avatar_data.get('name', 'unknown')}")

    # 2. Connect SDK session
    session = await client.sdk_connect(155, save_franchise_memory=True)
    print(f"SDK connected: {session}")

    # 3. Chat
    def on_msg(msg_type, text):
        print(f"  [{msg_type}] {text[:100]}")

    await client.chat(155, [
        "안녕! 오늘 기분 어때?",
        "요즘 영어 공부 하고 있는데 잘 되고 있는 거 같아?",
    ], on_response=on_msg)

    # 4. Disconnect
    await client.sdk_disconnect(155)
    print("Done!")


if __name__ == "__main__":
    asyncio.run(main())
